src/machine.rs

```diff
@@ -383,19 +401,44 @@ impl<'a, Inner: SupportMachine> DefaultMachine<'a, Inner> {
     pub fn interpret(&mut self) -> Result<u8, Error> {
         let decoder = build_imac_decoder::<Inner::REG>();
         self.set_running(true);
+        let mut traces: Vec<Trace> = Vec::new();
+        // For current trace size this is acceptable, however we might want
+        // to tweak the code here if we choose to use a larger trace size or
+        // larger trace item length.
+        traces.resize_with(TRACE_SIZE, Trace::default);
         while self.running() {
-            let instruction = {
-                let pc = self.pc().to_usize();
-                let memory = self.memory_mut();
-                decoder.decode(memory, pc)?
-            };
-            instruction.execute(self)?;
-            let cycles = self
-                .instruction_cycle_func()
-                .as_ref()
-                .map(|f| f(&instruction))
-                .unwrap_or(0);
-            self.add_cycles(cycles)?;
+            let pc = self.pc().to_usize();
+            let slot = (pc >> TRACE_ADDRESS_SHIFTS) & TRACE_MASK;
+            if pc != traces[slot].address {
+                for i in 0..TRACE_ITEM_LENGTH {
+                    traces[slot].instructions[i] = Instruction::Empty;
+                }
+                let mut current_pc = pc;
+                let mut i = 0;
+                while i < TRACE_ITEM_LENGTH {
+                    let instruction = decoder.decode(self.memory_mut(), current_pc)?;
+                    let end_instruction = is_basic_block_end_instruction(&instruction);
+                    current_pc += instruction_length(&instruction);
+                    traces[slot].instructions[i] = instruction;
+                    if end_instruction {
+                        break;
+                    }
+                    i += 1;
+                }
+                traces[slot].address = pc;
+            }
+            for i in &traces[slot].instructions {
+                if let Instruction::Empty = i {
```

Comment:
If Empty is only used as a placeholder here, I would suggest just use Option<Instruction> in the trace.